internal/era/era.go

```diff
@@ -0,0 +1,370 @@
+				if epoch, err := strconv.ParseUint(n[1], 10, 64); err != nil {
+					return nil, fmt.Errorf("malformed era1 filename: %s", entry.Name())
+				} else if epoch != next {
+					return nil, fmt.Errorf("missing epoch %d", next)
+				}
+				next += 1
+				eras = append(eras, entry.Name())
+			}
+		}
+	}
+	return eras, nil
+}
+
+// Reader reads an Era1 archive.
+// See Builder documentation for a detailed explanation of the Era1 format.
+type Reader struct {
+	r io.ReaderAt
+	e *e2store.Reader
+
+	buf      [8]byte  // buffer reading entry offsets
+	next     uint64   // next block to read
+	length   int64    // total length of r
+	metadata metadata // start, count info
+}
+
+type ReadAtSeeker interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// NewReader returns a new Reader instance.
+func NewReader(r ReadAtSeeker) (*Reader, error) {
+	length, err := r.Seek(0, io.SeekEnd)
+	if err != nil {
+		return nil, err
+	}
+	m, err := readMetadata(r, length)
+	if err != nil {
+		return nil, err
+	}
+	return &Reader{
+		r:        r,
+		e:        e2store.NewReader(r),
+		buf:      [8]byte{},
+		next:     m.start,
+		length:   length,
+		metadata: m,
+	}, nil
+}
+
+// readOffset reads a specific block's offset from the block index. The value n
+// is the absolute block number desired.
+func (r *Reader) readOffset(n uint64) (int64, error) {
+	var (
+		firstIndex  = -8 - int64(r.metadata.count)*8      // size of count - index entries
+		indexOffset = int64(n-r.metadata.start) * 8       // desired index * size of indexes
+		offOffset   = r.length + firstIndex + indexOffset // offset of block offset
+	)
+	r.clearBuffer()
+	if _, err := r.r.ReadAt(r.buf[:], offOffset); err != nil {
+		return 0, err
+	}
+	// Since the block offset is relative from its location + size of index
+	// value (8), we need to add it to it's offset to get the block's
+	// absolute offset.
+	return offOffset + 8 + int64(binary.LittleEndian.Uint64(r.buf[:])), nil
+}
+
+// Read reads one (block, receipts) tuple from an Era1 archive.
+func (r *Reader) Read() (*types.Block, types.Receipts, error) {
+	block, receipts, err := r.ReadBlockAndReceipts(r.next)
+	if err != nil {
+		return nil, nil, err
+	}
+	r.next += 1
+	return block, receipts, nil
+}
+
+// ReadHeader reads the header number n RLP.
+func (r *Reader) ReadHeaderRLP(n uint64) ([]byte, error) {
+	if n < r.metadata.start || r.metadata.start+r.metadata.count < n {
+		return nil, fmt.Errorf("request out-of-bounds: want %d, start: %d, count: %d", n, r.metadata.start, r.metadata.count)
+	}
+	// Read the specified block's offset from the block index.
+	off, err := r.readOffset(n)
+	if err != nil {
+		return nil, fmt.Errorf("error reading block offset: %w", err)
+	}
+	// Read header.
+	var e e2store.Entry
+	if _, err := r.e.ReadAt(&e, off); err != nil {
+		return nil, err
+	}
+	if e.Type != TypeCompressedHeader {
+		return nil, fmt.Errorf("expected header entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
```

Comment:
Since you will read the entire thing via io.ReadAll, why not just use 
```suggestion
	return snappy.Decode(nil, e.Value)
```

Reply:
Ah, it seems that you are encodign this using the snappy stream format, not the snappy block format. Is that a deliberate decision?

Reply:
Looking at how you implemented `snappyWrite`, it seems a bit like you are using the stream-format but manipulating it into block-format, using `Reset`  to restart new streams every time you might otherwise have used a block: 
```
	var (
		buf = b.buf
		s   = b.snappy
	)
	buf.Reset()
	s.Reset(buf)
```

Reply:
TLDR; I think this would be simpler if you just use `snappy.Encode` and `snappy.Decode`

Reply:
Yes exactly. So the [regular Era format](https://github.com/status-im/nimbus-eth2/blob/613f4a9a50c9c4bd8568844eaffb3ac15d067e56/docs/e2store.md#why-snappy-sz-framed-compression) is specified to use snappy stream encoding. On the beacon chain this makes sense, because it is the [same format ](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#ssz-snappy-encoding-strategy)they use for p2p.

I naively followed suit. Now that you bring it up though, it doesn't seem there is much benefit to use snappy stream for us? Our p2p is snappy block. At the same time, Era1 isn't exactly optimized for p2p requests. Post-merge Era files are also snappy stream encoded anyways, so it doesn't save us too much.

Reply:
>  If you receive 100 blocks that are frame (stream) -compressed,

Each block in the response is individually stream-compressed which among other things allows capping the amount of memory needed to decompress them. Ditto era files: all compression is done per "entry" so that we can retain O(1) access to all data. Framed encoding puts a cap on how much memory is needed for compression/decompression - we made a mistake on the CL side using block encoding at all (it's used in gossip, unfortunately).

Frame encoding also comes with a mime type and CRC32 which allows detecting simple corruption trivially - I'd strongly prefer that the eth1 format follows suite here and uses frame encoding, if not only for the simple reason that it makes life easier to have the same guarantees in both formats.